### PR TITLE
Add hardware signing and key exporting

### PR DESCRIPTION
**Pull request by AndrzejKurek**
This is a continuation of https://github.com/ARMmbed/mbed-os-atecc608a/pull/2.

This PR implements hardware ECDSA signing and public key exporting using ATECC608A or ATECC508A.
Prerequisite - a cryptoauth device (508A or 608A) comissioned, with a locked configuration and a private key generated in slot 0 (warning - configuration locking is irreversible!). An example application doing this with a quite safe config can be seen here.

This PR is a prerequisite for ARMmbed/mbed-os-example-atecc608a#2.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Why is this a separate PR from #2?

**Comment by AndrzejKurek** [comment]
@gilles-peskine-arm copied from #2:
>unfortunately due to the change of repository visibility (from private to public), my fork has been disconnected and this PR had to be recreated here: #3

Apparently, you can't change the PR's "from" part, and the PR didn't see any new commits added to the repository.

File: atecc608a_se.c [hunk 1]
```diff
@@ -15,6 +15,37 @@
...
+    do                    \
+    {                     \
+        atcab_release();  \
+    } while(0)
+
+#define ASSERT_STATUS(expression, expected, result)             \
+    do                                                          \
+    {                                                           \
+        status = (expression);                                  \
+        int ASSERT_STATUS_expected = (expected);                \
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c]: `status` is an `int32_t` storing PSA errors. `expected` is an `int` storing ATCA errors. You can't meaningfully compare the two, so `status = (expression)` doesn't make sense.

File: atecc608a_se.c [hunk 2]
```diff
@@ -77,52 +108,38 @@ static psa_status_t atecc608a_to_psa_error(ATCA_STATUS ret)
...
+    ATCAB_INIT();
 
-    if (atcab_release() != ATCA_SUCCESS)
-    {
-        return PSA_ERROR_HARDWARE_FAILURE;
-    }
+    ASSERT_SUCCESS(atcab_read_serial_number(buffer), PSA_ERROR_HARDWARE_FAILURE);
 
-    return PSA_SUCCESS;
+    exit:
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c]: indentation

File: atecc608a_se.c [hunk 3]
```diff
@@ -131,37 +148,29 @@ psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
         return PSA_ERROR_BUFFER_TOO_SMALL;
     }
 
-    if (atcab_init(&atca_iface_config) != ATCA_SUCCESS)
-    {
-        return PSA_ERROR_HARDWARE_FAILURE;
-    }
+    ATCAB_INIT();
+
+    ASSERT_SUCCESS((ret = atcab_get_pubkey(slot, &p_data[1])), atecc608a_to_psa_error(ret));
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c]: Suggestion: make a macro that embeds the `atecc608a_to_psa_error` transformation.

File: atecc608a_se.c [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        {                                                   \
+            status = PSA_ERROR_HARDWARE_FAILURE;            \
+            goto exit;                                      \
+        }                                                   \
+    } while(0)
+
+#define ATCAB_DEINIT()    \
+    do                    \
+    {                     \
+        atcab_release();  \
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c]: Realistically, what errors might `atcab_release` return? If it fails, should we report this as a `HARDWARE_FAILURE`?

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Looks good except for some confusion between ATCA and PSA return codes in the error-catching macros.

File: atecc608a_se.c (hunk 4 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: It might return ATCA_BAD_PARAM if there's no global device initialized, and an implementation of hal might return an error if an i2c device cannot be released. The implementations that I could find never return anything else than success.

File: atecc608a_se.c (hunk 1 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: I've made a mistake while writing this - please see the new version. Thanks!

**Comment by AndrzejKurek** [comment]
@gilles-peskine-arm - I've added the changes you've requested. Thanks!

File: atecc608a_se.c [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        ATCA_STATUS ASSERT_expected = (expected);               \
+        if ((ASSERT_result) != (ASSERT_expected))               \
+        {                                                       \
+            status = (psa_error);                               \
+            goto exit;                                          \
+        }                                                       \
+        status = PSA_SUCCESS;                                   \
+    } while(0)
+
+#define ASSERT_SUCCESS(expression) ASSERT_STATUS(expression,ATCA_SUCCESS, \
```
Comment (Patater) [inline atecc608a_se.c]: Style: I'd recommend spaces after commas for readabilty.

File: atecc608a_se.c [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    case ATCA_GEN_FAIL:
+    case ATCA_EXECUTION_ERROR:
+    case ATCA_HEALTH_TEST_ERROR:
+    case ATCA_INVALID_SIZE:
+    default:
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+}
+
+psa_status_t atecc608a_get_serial_number(uint8_t* buffer, size_t buffer_size,
```
Comment (Patater) [inline atecc608a_se.c]: Why add a serial number getting function as a public function? Seems unrelated to implementing signature and key exporting, at least. This is potentially useful for verifying that the hardware is reachable and returns something when asked, but I don't see us needing this right now; it's just a fun debug print in an example.

File: atecc608a_se.c [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+    ASSERT_SUCCESS(atcab_read_serial_number(buffer));
+    *buffer_length = ATCA_SERIAL_NUM_SIZE;
+
+exit:
+    ATCAB_DEINIT();
+    return status;
+}
+
+psa_status_t atecc608a_check_config_locked()
```
Comment (Patater) [inline atecc608a_se.c]: Can this be static?

File: atecc608a_se.c [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+exit:
+    ATCAB_DEINIT();
+    if(status == PSA_SUCCESS)
+    {
+        status = config_locked? PSA_SUCCESS : PSA_ERROR_HARDWARE_FAILURE;
+    }
+    return status;        
+}
+
+psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
```
Comment (Patater) [inline atecc608a_se.c]: Can this be static?

File: atecc608a_se.c [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    printf("atecc608a_export_key - pubkey size %d:\n", *p_data_length);
+    atcab_printbin_sp(p_data, *p_data_length);
+#endif
+
+exit:
+    ATCAB_DEINIT();
+    return status;
+}
+
+psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
```
Comment (Patater) [inline atecc608a_se.c]: Can this be static?

File: atecc608a_se.c [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+#ifdef DEBUG_PRINT
+    printf("atecc608a_asymmetric_sign - signature size %d:\n", *p_signature_length);
+    atcab_printbin_sp(p_signature, *p_signature_length);
+#endif
+
+exit:
+    ATCAB_DEINIT();
+    return status;
+}
```
Comment (Patater) [inline atecc608a_se.c:209]: To make these functions static, the driver C file should have a global, linker-visible struct filled in with the driver callbacks. The example or anything using the atecc608s can then use the functions through the struct.

File: atecc608a_se.h [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+psa_status_t atecc608a_check_config_locked();
+psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key, uint8_t *p_data,
+                                         size_t data_size, size_t *p_data_length);
+psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
+                                       psa_algorithm_t alg,
+                                       const uint8_t *p_hash,
+                                       size_t hash_length,
+                                       uint8_t *p_signature,
+                                       size_t signature_size,
+                                       size_t *p_signature_length);
```
Comment (Patater) [inline atecc608a_se.h]: I'd recommend only exporting the driver struct from the header file, keeping all implementation internal to the C file.

**Review by Patater — CHANGES_REQUESTED** [review]

File: atecc608a_se.c (hunk 10 above)
Comment (AndrzejKurek) [reply atecc608a_se.c:209]: Sure, we can do it this way if it's the only thing we want to have here. I've even thought about adding a function to print the configuration zone (this might be useful for validation), but after your suggestion - we can have that in an utils.h in the example too.

File: atecc608a_se.c (hunk 4 above)
Comment (gilles-peskine-arm) [reply atecc608a_se.c]: Ok. Please add this information in a comment somewhere to justify ignoring errors from `atcab_release`.

File: atecc608a_se.c [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        }                                                   \
+    } while(0)
+
+#define ATCAB_DEINIT()    \
+    do                    \
+    {                     \
+        atcab_release();  \
+    } while(0)
+
+#define ASSERT_STATUS(expression, expected, psa_error)          \
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c:39]: Please add a comment that explains briefly what the macro does and how to use it.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
The code looks good, a couple of places need comments.

**Comment by AndrzejKurek** [comment]
@Patater, @gilles-peskine-arm - I've added your suggestions. Please re-review.

**Review by gilles-peskine-arm — APPROVED** [review]

**Comment by AndrzejKurek** [comment]
@gilles-peskine-arm - I've made the utility functions and macros that you've mentioned visible.

File: atecc608a_se.h [hunk 13]
```diff
@@ -23,8 +23,33 @@
 #ifndef ATECC608A_SE_H
 #define ATECC608A_SE_H
 
-#include "psa/crypto_se_driver.h"
+#include "psa/crypto.h"
```
Comment (gilles-peskine-arm) [inline atecc608a_se.h]: You shouldn't need to include `crypto.h` here. Why isn't `crypto_se_driver.h` sufficient? If it's missing something, that's probably a bug in that header.

Comment (AndrzejKurek) [reply atecc608a_se.h]: amended.

File: atecc608a_se.c [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+            goto exit;                                          \
+        }                                                       \
+        status = PSA_SUCCESS;                                   \
+    } while(0)
+
+/** Check if an ATCA operation is successful, translate the error otherwise. */
+#define ASSERT_SUCCESS(expression) ASSERT_STATUS(expression, ATCA_SUCCESS, \
+                                      atecc608a_to_psa_error(ASSERT_result))
+
+ATCAIfaceCfg atca_iface_config = {
```
Comment (Patater) [inline atecc608a_se.c]: This can be `const`, to same some RAM.

I also see this is public. Why not `static`? What use is it outside the driver? I guess if folks want to use the same configuration for the SE that the driver is using, they could use this. It's only going to be useful for PSA Services and not applications (like our example), though.

File: atecc608a_se.c [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+    .iface_type = ATCA_I2C_IFACE,
+    .devtype = ATECC608A,
+    .atcai2c.slave_address = 0xC0,
+    .atcai2c.bus = 2,
+    .atcai2c.baud = 400000,
+    .wake_delay = 1500,
+    .rx_retries = 20,
+};
+
+psa_status_t atecc608a_to_psa_error(ATCA_STATUS ret)
```
Comment (Patater) [inline atecc608a_se.c:71]: What use this this outside the driver? Why shouldn't this be static?

If an application is working with PSA error types, I'd expect it to always use those types and not be fiddling with ATCA statuses. A PSA Service might find this useful, potentially.

File: atecc608a_se.c [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+    atcab_printbin_sp(p_signature, *p_signature_length);
+#endif
+
+exit:
+    ATCAB_DEINIT();
+    return status;
+}
+
+
+#define PSA_ATECC608A_LIFETIME 0xdeadbeefU
```
Comment (Patater) [inline atecc608a_se.c:212]: This should probably be in the header file, so that applications can ask for a slot from the SE using it.

File: atecc608a_se.c [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+    .p_encrypt = 0,
+    .p_decrypt = 0,
+};
+
+static psa_drv_se_key_management_t atecc608a_key_management =
+{
+    .p_import = 0,
+    .p_generate = 0,
+    .p_destroy = 0,
+    /* So far there is no public key export function in the API, so use this instead */
```
Comment (Patater) [inline atecc608a_se.c:227]: Yes, we need to fix that, outside the scope of this PR.

**Review by Patater — CHANGES_REQUESTED** [review]

File: atecc608a_se.c (hunk 15 above)
Comment (AndrzejKurek) [reply atecc608a_se.c:71]: This can be used by all other utility functions that we write for this secure element that do not fit the driver - like provisioning, config printing, misc. operations.

File: atecc608a_se.c (hunk 16 above)
Comment (AndrzejKurek) [reply atecc608a_se.c:212]: This can be changed later on, when full (or just better) support of the SE is added.

File: atecc608a_se.c (hunk 14 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: This cannot be const, because `atcab_init` requires a non-const object, and parts the object itself are probably modified by the implementation (there is, for example an opaque configuration data pointer that we shouldn't interfere with).
This is not static because of the `ATCAB_INIT` macro - each use of it uses `atca_iface_config`. `ATCAB_INIT` is used by our sample app, but I can imagine other code using it too. For non-standard hardware configurations however, I think that we should consider injecting the interface config into the SE configuration.

Comment (Patater) [reply atecc608a_se.c]: OK, was not aware that this needed to be writable. That's unfortunate.

I disagree that our example application should talk to the SE directly. OK for development for now, but eventually it should use only PSA APIs and not talk to the SE directly.

**Review by Patater — DISMISSED** [review]

File: atecc608a_se.c (hunk 14 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: What about provisioning and miscellaneous operations, like config/serial number printing? I'm not sure if we'll ever have a place for that, and still it's useful. I've duplicated the code in my first approach, but Gilles deemed it a bigger issue.

**Review by gilles-peskine-arm — DISMISSED** [review]

File: atecc608a_se.c (hunk 14 above)
Comment (Patater) [reply atecc608a_se.c]: It's a big design issue we haven't addressed yet. We do need a way to provision SEs, but how that can be done isn't designed yet.

Comment (gilles-peskine-arm) [reply atecc608a_se.c]: There are two ways to do provisioning. One is to call `psa_import_key`, and it's already fully supported in the interface. The other is that keys are provisioned by bypassing the normal API (typically, they appear “by magic” during manufacturing), and we're still working on how to declare such keys so that they can be used via PSA.

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by Patater — APPROVED** [review]
